### PR TITLE
[WNMGDS-2818] Prevent nested errors in choicelist

### PR DESCRIPTION
**Pull request by zarahzachz**
WNMGDS-2818

Prevent errors from being nested in ChoiceList.

- Individual Choice can have an `errorMessage`
- Choices nested within ChoiceList cannot have an `errorMessage`
- If an error exists in ChoiceList, it should only be revealed by the ChoiceList's `errorMessage`, not on its children

Also added TS checks to make sure `error*` props weren't being used and a `console.warn` to gently scold if they weren't TS users and insisted on doing the bad pattern.

### CODE REVIEW

File: packages/design-system/src/components/ChoiceList/Choice.tsx [hunk 1]
```diff
@@ -112,13 +116,16 @@ export const Choice = (props: ChoiceProps) => {
...
   let errorElement;
-  if (props.errorMessage) {
+  if (!props._choiceChild) {
     errorId = props.errorId ?? `${props.id}__error`;
     errorElement = (
       <InlineError id={errorId} inversed={props.inversed} className={props.errorMessageClassName}>
         {props.errorMessage}
       </InlineError>
     );
+  } else {
```
Comment (zarahzachz) [inline packages/design-system/src/components/ChoiceList/Choice.tsx:126]: This condition needed to be added to strip `_choiceChild` from rendered markup (failed snapshot tests).

File: packages/design-system/src/components/ChoiceList/ChoiceList.tsx [hunk 2]
```diff
@@ -19,7 +19,10 @@ export interface BaseChoiceListProps {
   /**
    * Array of objects representing the props for each Choice in the ChoiceList
    */
-  choices: Omit<ChoiceProps, 'name' | 'type'>[];
+  choices: Omit<
+    ChoiceProps,
+    'name' | 'type' | 'errorMessage' | 'errorId' | 'errorMessageClassName'
+  >[];
```
Comment (pwolfert) [inline packages/design-system/src/components/ChoiceList/ChoiceList.tsx:25]: 👍

File: packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap [hunk 3]
```diff
@@ -132,6 +139,12 @@ exports[`Choice is a checkbox 1`] = `
         >
           George Washington
         </label>
+        <p
+          aria-atomic="true"
+          aria-live="assertive"
+          class="ds-c-inline-error"
+          id="undefined__error"
```
Comment (pwolfert) [inline packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap:146]: This is going to be duplicated in the DOM pretty quickly, so I don't think we can ship it in this state. Do you know why this is undefined?

Comment (pwolfert) [reply packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap:146]: Looks like it's actually happening on the React Choice in Storybook too

File: packages/design-system/src/components/ChoiceList/Choice.tsx [hunk 4]
```diff
@@ -112,13 +116,16 @@ export const Choice = (props: ChoiceProps) => {
...
+    const { _choiceChild, ...rest } = props;
+    props = rest;
```
Comment (pwolfert) [inline packages/design-system/src/components/ChoiceList/Choice.tsx:128]: What's the advantage of this over `delete props._choiceChild`? It's still mutating the pointer `props` even if it doesn't mutate the object. I think delete is a little clearer

File: packages/design-system/src/components/ChoiceList/Choice.tsx [hunk 5]
```diff
@@ -112,13 +116,16 @@ export const Choice = (props: ChoiceProps) => {
 
   let errorId;
   let errorElement;
-  if (props.errorMessage) {
+  if (!props._choiceChild) {
     errorId = props.errorId ?? `${props.id}__error`;
```
Comment (pwolfert) [inline packages/design-system/src/components/ChoiceList/Choice.tsx:120]: I think this assignment statement being in a different kind of condition now might be one clue about what's happening

File: packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap (hunk 3 above)
Comment (zarahzachz) [reply packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap:146]: This is some thing I was seeing in PROD right now. I think you can see it in the web component side of things.

File: packages/design-system/src/components/ChoiceList/Choice.tsx (hunk 4 above)
Comment (zarahzachz) [reply packages/design-system/src/components/ChoiceList/Choice.tsx:128]: Using delete caused tests for MonthPicker and ChoiceList to fail. I can't remember what the error was, I think it had to do with a type error where you couldn't delete something from an object.

Comment (pwolfert) [reply packages/design-system/src/components/ChoiceList/Choice.tsx:128]: Huh, it's kind of odd that TypeScript would be mad about that but fine with reassigning `rest` to `props` even though it's a slightly different type. If we moved it up to the function signature like

```ts
export const Choice = ({ _choiceChild, ...props }: ChoiceProps) => {
```

then we wouldn't need the `else` block at all.

File: packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap (hunk 3 above)
Comment (pwolfert) [reply packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap:146]: Okay, I see it in prod now

Comment (pwolfert) [reply packages/design-system/src/components/web-components/ds-choice/__snapshots__/ds-choice.test.tsx.snap:146]: [Created a fix for it](https://github.com/CMSgov/design-system/pull/3146)

**Review by pwolfert — APPROVED** [review]
LGTM 👍

File: packages/design-system/src/components/ChoiceList/Choice.tsx [hunk 6]
```diff
@@ -44,6 +44,10 @@ export interface BaseChoiceProps {
    * Additional classes to be added to the root `div` element.
    */
   className?: string;
+  /**
+   * @hide-prop Internal prop used to determine if a Choice is the child of a another component (like ChoiceList or MonthPicker). Used to hide excessive error messages.
+   */
+  _choiceChild?: boolean;
```
Comment (pwolfert) [inline packages/design-system/src/components/ChoiceList/Choice.tsx:50]: Right now the doc snapshots are failing, and I don't think we have a working method of hiding any props from the arg tables. We can do it for web components where we created a custom table component, but I don't know how to hide it from the arg tables in the storybook one. The following will only hide it from the storybook controls but not the doc page:

```js
argTypes: {
  _choiceChild: {
    // Until this pattern has solidified, we're not going to advertize this feature.
    table: {
      disable: true,
    },
  },
}
```

**Pull request merged**